File: frontend/src/lib/api/qa.ts
/**
 * QA, cost, audit, thumbnail and command endpoints.
 *
 * Covers the router built from ``models/qa.py``. Two things worth knowing:
 *
 * * Each check has **two** routes: a list form (``/qa/platform``) returning raw
 *   findings and a verdict form (``/qa/platform/verdict``) returning a pass/fail
 *   summary. The studio uses the verdict form and shows the findings it carries.
 * * ``/cost/check`` merges ``calls`` and ``estimated_usage``; the studio sends
 *   ``estimated_usage``, which is why :interface:`CostCheckRequest` lists both.
 */

import * as api from "@/lib/api/client";
import {
  AuditEntry,
  AuditRecordRequest,
  BrandCheckRequest,
  BrandVerdict,
  CopyrightCheckRequest,
  CopyrightVerdict,
  CostCheckRequest,
  CostCheckResult,
  PlatformCheckRequest,
  PlatformVerdict,
  QaFinding,
  SimplifySubtitlesRequest,
  SimplifySubtitlesResult,
  ThumbnailCandidatesResponse,
  ThumbnailRequest,
} from "@/types/qa";

/** ``POST /qa/platform`` — the raw findings. */
export function checkPlatform(req: PlatformCheckRequest): Promise<QaFinding[]> {
  return api.apiFetch<QaFinding[]>("/qa/platform", { method: "POST", body: req });
}

/** ``POST /qa/platform/verdict`` — pass/fail plus the findings behind it. */
export function platformVerdict(
  req: PlatformCheckRequest
): Promise<PlatformVerdict> {
  return api.apiFetch<PlatformVerdict>("/qa/platform/verdict", {
    method: "POST",
    body: req,
  });
}

/** ``POST /qa/brand`` — the raw findings. */
export function checkBrand(req: BrandCheckRequest): Promise<QaFinding[]> {
  return api.apiFetch<QaFinding[]>("/qa/brand", { method: "POST", body: req });
}

/** ``POST /qa/brand/verdict``. */
export function brandVerdict(req: BrandCheckRequest): Promise<BrandVerdict> {
  return api.apiFetch<BrandVerdict>("/qa/brand/verdict", {
    method: "POST",
    body: req,
  });
}

/** ``POST /qa/copyright`` — the raw findings. */
export function checkCopyright(req: CopyrightCheckRequest): Promise<QaFinding[]> {
  return api.apiFetch<QaFinding[]>("/qa/copyright", { method: "POST", body: req });
}

/** ``POST /qa/copyright/verdict``. */
export function copyrightVerdict(
  req: CopyrightCheckRequest
): Promise<CopyrightVerdict> {
  return api.apiFetch<CopyrightVerdict>("/qa/copyright/verdict", {
    method: "POST",
    body: req,
  });
}

/** ``GET /audit`` — the newest audit records first. */
export function listAudit(limit = 100): Promise<AuditEntry[]> {
  return api.apiFetch<AuditEntry[]>("/audit", { query: { limit } });
}

/** ``POST /audit/record`` — append one provenance record. */
export function recordAudit(req: AuditRecordRequest): Promise<AuditEntry> {
  return api.apiFetch<AuditEntry>("/audit/record", { method: "POST", body: req });
}

/** ``POST /cost/check`` — price a plan of model calls. */
export function checkCost(req: CostCheckRequest): Promise<CostCheckResult> {
  return api.apiFetch<CostCheckResult>("/cost/check", {
    method: "POST",
    body: req,
  });
}

/**
 * ``POST /thumbnail/candidates``.
 *
 * Rejects with :class:`ApiError` only on transport/HTTP failure; "nothing to
 * draw from" comes back as ``empty_reason`` with an empty list, so the caller can
 * explain the empty grid rather than showing a bare error.
 */
export function thumbnailCandidates(
  req: ThumbnailRequest
): Promise<ThumbnailCandidatesResponse> {
  return api.apiFetch<ThumbnailCandidatesResponse>("/thumbnail/candidates", {
    method: "POST",
    body: req,
  });
}

/** ``POST /subtitles/simplify`` — readable captions for cognitive accessibility. */
export function simplifySubtitles(
  req: SimplifySubtitlesRequest
): Promise<SimplifySubtitlesResult> {
  return api.apiFetch<SimplifySubtitlesResult>("/subtitles/simplify", {
    method: "POST",
    body: req,
  });
}

/** ``GET /health`` — status, version and the live provider chain. */
export function health(): Promise<Record<string, unknown>> {
  return api.apiFetch<Record<string, unknown>>("/health");
}

File: src/content_factory/api/routers/qa.py
"""QA, traceability, media-intelligence and production-booster endpoints.

Exposes the new additive modules over HTTP so the web app and external agents
can use them without touching the core pipeline:

  * QA / risk  -> `compliance.py` (platform, brand, copyright)
  * Traceability -> `audit.py` (provenance log) + `cost_guard.py` (cost gate)
  * Media intelligence -> `dedup.py` + `search.py`
  * Production boosters -> `virality.py`, `audio.py` (ducking), `thumbnail.py`

All of these are deterministic/offline; none bypass the two human gates.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter

from ...models import (
    AuditRecordRequest,
    BrandCheckRequest,
    CopyrightCheckRequest,
    CostCheckRequest,
    DedupRequest,
    DuckRequest,
    PlatformCheckRequest,
    SimplifySubtitlesRequest,
    ThumbnailRequest,
    TimelineCommandRequest,
    ViralityRequest,
)
from ...service import ContentFactoryService
from ..deps import guard_value


def build_router(service: ContentFactoryService) -> APIRouter:
    router = APIRouter()

    @router.post("/qa/platform")
    def qa_platform(req: PlatformCheckRequest) -> list[dict[str, Any]]:
        """Review a video against one platform's rules (duration/aspect/words)."""
        return service.qa_platform(req)

    @router.post("/qa/platform/verdict")
    def qa_platform_verdict(req: PlatformCheckRequest) -> dict[str, Any]:
        """The same platform check as a verdict: pass/fail, messages, advice."""
        return service.qa_platform_verdict(req)

    @router.post("/qa/brand")
    def qa_brand(req: BrandCheckRequest) -> list[dict[str, Any]]:
        """Check the video against a brand kit (palette / logo / fonts)."""
        return service.qa_brand(req)

    @router.post("/qa/brand/verdict")
    def qa_brand_verdict(req: BrandCheckRequest) -> dict[str, Any]:
        """The same brand check as a verdict, with categorised findings."""
        return service.qa_brand_verdict(req)

    @router.post("/qa/copyright")
    def qa_copyright(req: CopyrightCheckRequest) -> list[dict[str, Any]]:
        """Flag an exact fingerprint match against a protected set.

        Accepts a single ``fingerprint`` or a batch of ``asset_ids``.
        """
        return service.qa_copyright(req)

    @router.post("/qa/copyright/verdict")
    def qa_copyright_verdict(req: CopyrightCheckRequest) -> dict[str, Any]:
        """The same copyright check as a verdict, one row per checked asset."""
        return service.qa_copyright_verdict(req)

    @router.get("/audit")
    def audit_list(limit: int = 100) -> list[dict[str, Any]]:
        """Recent provenance entries (newest first)."""
        return service.audit_list(limit)

    @router.post("/audit/record")
    def audit_record(req: AuditRecordRequest) -> dict[str, Any]:
        """Append one provenance entry (which model, what action, when)."""
        return service.audit_record(req)

    @router.post("/cost/check")
    def cost_check(req: CostCheckRequest) -> dict[str, Any]:
        """Estimate the USD cost of an expensive plan and whether to confirm."""
        return service.cost_check(req)

    @router.get("/cost/estimate")
    def cost_estimate(
        vision: int = 0,
        audio_llm: int = 0,
        tts: int = 0,
        stt: int = 0,
        embedding: int = 0,
    ) -> dict[str, Any]:
        """The same estimate from query parameters, so a UI can poll it."""
        return service.cost_check(
            CostCheckRequest(
                calls={
                    "vision": vision,
                    "audio_llm": audio_llm,
                    "tts": tts,
                    "stt": stt,
                    "embedding": embedding,
                }
            )
        )

    @router.post("/media/dedup")
    def media_dedup(req: DedupRequest) -> list[list[str]]:
        """Group media items whose perceptual hashes are near-duplicates."""
        return guard_value(lambda: service.media_dedup(req))

    @router.get("/media/search")
    def media_search(q: str, top_k: int = 10) -> list[dict[str, Any]]:
        """Search the media library by transcript/filename (lexical + embedder)."""
        return service.media_search(q, top_k=top_k)

    @router.post("/script/virality")
    def script_virality(req: ViralityRequest) -> dict[str, Any]:
        """Heuristic pre-publish virality score (hook / pace / length / CTA)."""
        return service.script_virality(req)

    @router.post("/render/duck")
    def render_duck(req: DuckRequest) -> dict[str, Any]:
        """Duck a music bed under a voice track (ffmpeg sidechaincompress)."""
        return guard_value(lambda: service.render_duck(req))

    @router.post("/thumbnail/generate")
    def thumbnail_generate(req: ThumbnailRequest) -> list[dict[str, Any]]:
        """Generate thumbnail candidates from the best frames + CTR prediction.

        Returns the bare list (the dashboard and the agent tools iterate it).
        ``POST /thumbnail/candidates`` wraps the same list for clients that
        expect a ``{candidates}`` envelope.
        """
        return guard_value(lambda: service.thumbnail_generate(req))

    @router.post("/thumbnail/candidates")
    def thumbnail_candidates(req: ThumbnailRequest) -> dict[str, Any]:
        """The same candidates, wrapped as ``{"candidates": [...], "count": n}``.

        Carries ``empty_reason``/``emptyReason`` when nothing was drawable, so a
        client can say *why* instead of showing an unexplained empty grid.
        """
        candidates = guard_value(lambda: service.thumbnail_generate(req))
        payload: dict[str, Any] = {
            "candidates": candidates,
            "count": len(candidates),
        }
        if not candidates:
            reason = service.thumbnail_empty_reason(req)
            payload["empty_reason"] = reason
            payload["emptyReason"] = reason
        return payload

    @router.post("/timeline/command")
    def timeline_command(req: TimelineCommandRequest) -> dict[str, Any]:
        """Apply a natural-language editing instruction to a video project."""
        return service.timeline_command(req)

    @router.post("/subtitles/simplify")
    def subtitles_simplify(req: SimplifySubtitlesRequest) -> dict[str, Any]:
        """Simplify captions for kids / language learners (accessibility)."""
        return service.subtitles_simplify(req)

    return router
